pkg/controller/direct/discoveryengine/datastore_controller.go

```diff
@@ -181,34 +186,37 @@ func (a *dataStoreAdapter) Create(ctx context.Context, createOp *directbase.Crea
 
 	status := &krm.DiscoveryEngineDataStoreStatus{}
 	mapCtx := &direct.MapContext{}
-	status.ObservedState = DiscoveryEngineDataStoreObservedState_FromProto(mapCtx, created)
+	status.ObservedState = DataStoreObservedState_FromProto(mapCtx, created)
 	if mapCtx.Err() != nil {
 		return mapCtx.Err()
 	}
 	status.ExternalRef = direct.PtrTo(a.id.String())
+
 	return createOp.UpdateStatus(ctx, status, nil)
 }
 
 func (a *dataStoreAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
 	log := klog.FromContext(ctx)
 	log.V(2).Info("updating discoveryengine datastore", "name", a.id)
 
-	desired := direct.ProtoClone(a.desired)
-	desired.Name = a.id.String()
-
-	// TODO(user): Update the field if applicable.
-	updateMask := &fieldmaskpb.FieldMask{}
-	if !reflect.DeepEqual(a.desired.DisplayName, a.actual.DisplayName) {
-		updateMask.Paths = append(updateMask.Paths, "display_name")
+	a.desired.Name = a.actual.Name
+	paths := make(sets.Set[string])
+	var err error
+	paths, err = common.CompareProtoMessage(a.desired, a.actual, common.BasicDiff)
```

Comment:
Do you need to mask out so that a.actual is just the spec fields?  (One trick to do so is to convert it to spec and back!)

Reply:
We don't need to. CompareProtoMessage skips OUTOUT-ONLY fields. So the proto level comparison is spec to spec.